kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,759 @@
+    // Bug filed at https://youtrack.jetbrains.com/issue/KT-22783
+    val file = filer.createSourceFile(adapterName).toUri().let(::File)
+    return file.parentFile.also { file.delete() }
+  }
+}
+
+/**
+ * Creates a joined string representation of simplified typename names.
+ */
+private fun List<TypeName>.simplifiedNames(): String {
+  return joinToString("_") { it.simplifiedName() }
+}
+
+private fun TypeName.resolveRawType(): ClassName {
+  return when (this) {
+    is ClassName -> this
+    is ParameterizedTypeName -> rawType
+    else -> throw UnsupportedOperationException("Cannot get raw type from $this")
+  }
+}
+
+/**
+ * Creates a simplified string representation of a TypeName's name
+ */
+private fun TypeName.simplifiedName(): String {
+  return when (this) {
+    is ClassName -> simpleName().decapitalize()
+    is ParameterizedTypeName -> {
+      rawType.simpleName().decapitalize() + if (typeArguments.isEmpty()) "" else "__" + typeArguments.simplifiedNames()
+    }
+    is WildcardTypeName -> "wildcard__" + (lowerBounds + upperBounds).simplifiedNames()
+    is TypeVariableName -> name.decapitalize() + if (bounds.isEmpty()) "" else "__" + bounds.simplifiedNames()
+  // Shouldn't happen
+    else -> toString().decapitalize()
+  }.let { if (nullable) "${it}_nullable" else it }
+}
+
+private fun ClassName.isClass(elementUtils: Elements): Boolean {
+  val fqcn = toString()
+  if (fqcn.startsWith("kotlin.collections.")) {
+    // These are special kotlin interfaces are only visible in kotlin, because they're replaced by
+    // the compiler with concrete java classes
+    return false
+  } else if (this == ARRAY) {
+    // This is a "fake" class and not visible to Elements
+    return true
+  }
+  return elementUtils.getTypeElement(fqcn).kind == ElementKind.INTERFACE
+}
+
+private fun TypeName.objectType(): TypeName {
+  return when (this) {
+    BOOLEAN -> Boolean::class.javaObjectType.asTypeName()
+    BYTE -> Byte::class.javaObjectType.asTypeName()
+    SHORT -> Short::class.javaObjectType.asTypeName()
+    INT -> Integer::class.javaObjectType.asTypeName()
+    LONG -> Long::class.javaObjectType.asTypeName()
+    CHAR -> Character::class.javaObjectType.asTypeName()
+    FLOAT -> Float::class.javaObjectType.asTypeName()
+    DOUBLE -> Double::class.javaObjectType.asTypeName()
+    else -> this
+  }
+}
+
+private fun TypeName.makeType(
+    elementUtils: Elements,
+    typesArray: ParameterSpec,
+    genericTypeNames: List<TypeVariableName>): CodeBlock {
+  if (nullable) {
+    return asNonNullable().makeType(elementUtils, typesArray, genericTypeNames)
+  }
+  return when (this) {
+    is ClassName -> CodeBlock.of("%T::class.java", this)
+    is ParameterizedTypeName -> {
+      // If it's an Array type, we shortcut this to return Types.arrayOf()
+      if (rawType == ARRAY) {
+        return CodeBlock.of("%T.arrayOf(%L)",
+            Types::class.asTypeName(),
```

Comment:
General heads-up: KotlinPoet understands KClass and so you don't need to box instances when emitting to a `%T`

Reply:
Neat, I wasn't sure. Did some cleanup in 0390cc5